### PR TITLE
Fb dataset notifications 9061

### PR DESCRIPTION
**Pull request by labkey-klum**
#### Rationale
We want to expand the options available when configuring notifications for report and dataset changes in a study. With this change we now support selection on a per dataset basis which will be mutually exclusive from the other previous options:
- All
- By category

In addition the notification sent out will contain a link (for datasets) to the audit changes that occurred for that notification. 

#### Related Documents
- [Specification](https://docs.google.com/document/d/1_7mF64TIT5b5x28RSw3S1G8xrmj_avaJyrWGDi481Dk/edit)

#### Changes
- Add a new option to the notification config UI to set on a per dataset basis
- Refactor the Ext4 code to be more consistent with other components that we have written in Ext4. Extend the panel class instead of acting like a factory class.
- Update the email template and formatting (styles were not included properly)
- Simplify the HTML generation of the report and notification email template grid. Switch from HTML generation in java to rendering using a JSP.
- Add an audit link to the manage datasets page
- Collect metrics for the number of notification configurations on the server

### CODE REVIEW

File: study/src/org/labkey/study/controllers/DatasetController.java [hunk 1]
```diff
@@ -102,7 +103,7 @@ protected void addExtraNavTrail(NavTree root)
         }
     }
 
-    @RequiresPermission(AdminPermission.class)
+    @RequiresPermission(CanSeeAuditLogPermission.class)
```
Comment (labkey-klum) [inline study/src/org/labkey/study/controllers/DatasetController.java:107]: Checked with Susan, no reason that this view should be restricted to admin only.

File: query/src/org/labkey/query/reports/view/ReportAndDatasetChangeDigestEmailTemplate.java [hunk 2]
```diff
@@ -111,45 +112,36 @@ public void init(Container c, Map<ViewCategory, List<NotificationInfo>> reports)
...
+
+    private String buildReportAndDatasetList(Container c, Map<ViewCategory, List<NotificationInfo>> reports)
+    {
+        try
+        {
+            NotificationBean bean = new NotificationBean();
+            bean.setReports(reports);
 
-        return sb.toString();
+            ViewContext context = ViewContext.getMockViewContext(User.getSearchUser(), c, new ActionURL(), true);
```
Comment (labkey-adam) [inline query/src/org/labkey/query/reports/view/ReportAndDatasetChangeDigestEmailTemplate.java]: Can you use a `JspTemplate` here instead?

File: query/src/org/labkey/query/reports/view/manageNotifications.jsp [hunk 3]
```diff
@@ -33,62 +32,33 @@
...
-        var manager = Ext4.create('LABKEY.ext4.ManageReportNotifications');
-        var panel = manager.getManageReportPanel({
-                    renderTo : 'manageNotificationsDiv',
-                    minWidth : 750
-                },
-                categories, returnUrl, notifyOption);
+        Ext4.create('LABKEY.ext4.ReportNotificationPanel', {
+            title : 'Choose Notification Option',
+            categories : <%=HtmlString.unsafe(jsonMapper.writeValueAsString(form.getCategories()))%>,
+            datasets : <%=HtmlString.unsafe(jsonMapper.writeValueAsString(form.getDatasets()))%>,
```
Comment (labkey-adam) [inline query/src/org/labkey/query/reports/view/manageNotifications.jsp]: I think you could use `<%=toJsonArray()%>` or `<%=toJsonObject()%>` here. More succinct and their output is already marked `SafeToRender`.

**Review by labkey-adam — CHANGES_REQUESTED** [review]
Just a couple suggestions to consider

**Comment by labkey-adam** [comment]
Did you want to target `release21.7-SNAPSHOT`?

File: query/src/org/labkey/query/reports/view/manageNotifications.jsp (hunk 3 above)
Comment (labkey-klum) [reply query/src/org/labkey/query/reports/view/manageNotifications.jsp]: 👍

File: query/src/org/labkey/query/reports/view/ReportAndDatasetChangeDigestEmailTemplate.java (hunk 2 above)
Comment (labkey-klum) [reply query/src/org/labkey/query/reports/view/ReportAndDatasetChangeDigestEmailTemplate.java]: ah right, for the life of me I couldn't remember the name of that class.

File: query/src/org/labkey/query/reports/view/ReportAndDatasetChangeDigestEmailTemplate.java [hunk 4]
```diff
@@ -111,45 +112,36 @@ public void init(Container c, Map<ViewCategory, List<NotificationInfo>> reports)
...
+            JspTemplate<NotificationBean> view = new JspTemplate<>("/org/labkey/query/reports/view/reportAndDatasetList.jsp", bean);
```
Comment (labkey-adam) [inline query/src/org/labkey/query/reports/view/ReportAndDatasetChangeDigestEmailTemplate.java]:
```suggestion
            return new JspTemplate<>("/org/labkey/query/reports/view/reportAndDatasetList.jsp", bean).render();

```

Comment (labkey-adam) [reply query/src/org/labkey/query/reports/view/ReportAndDatasetChangeDigestEmailTemplate.java]: You don't need the `MockViewContext`, etc. -- `JspTemplate.render()` will handle it for you

**Review by labkey-adam — CHANGES_REQUESTED** [review]
Simplify `JspTemplate` handling

**Review by labkey-adam — APPROVED** [review]

**Pull request merged**